ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -33,77 +33,96 @@ def __init__(self):
+        self.pose= None
+        self.stopline_wp_idx= -1
+        self.waypoints_2d= None
+        self.waypoint_tree= None
+        #rospy.spin()
         self.loop()
 
-    # Looping at 50 Hz to get control over publishing frequency
     def loop(self):
-        # Rate should at least be >= 30 Hz
         rate = rospy.Rate(50)
         while not rospy.is_shutdown():
-            if self.pose and self.base_waypoints:
-                # Get closest waypoint
-                closest_waypoint_idx = self.get_closest_waypoint_idx()
-                self.publish_waypoints(closest_waypoint_idx)
+            if self.pose and self.base_lane:
+                self.publish_waypoints()
             rate.sleep()
 
     def get_closest_waypoint_idx(self):
         x = self.pose.pose.position.x
         y = self.pose.pose.position.y
-        # KDTree access is O(log(n)), versus O(n) for brute-force scanning all waypoints
-        closest_idx = self.waypoint_tree.query([x, y], 1)[1]
+        closest_idx = self.waypoint_tree.query([x,y],1)[1]
 
-        # Check if closest is ahead or behind vehicle? (want ones ahead only)
-        closest_coord = self.waypoints_2d[closest_idx]
-        prev_coord = self.waypoints_2d[closest_idx - 1]
+        #check if closest is ahead or behind vehicle
+        closest_coord= self.waypoints_2d[closest_idx]
+        prev_coord = self.waypoints_2d[closest_idx-1]
 
-        # Equation for hyperplane through closest_coords
+        #equation for hyperplane through closest_coords
         cl_vect = np.array(closest_coord)
         prev_vect = np.array(prev_coord)
-        pos_vect = np.array([x, y])
+        pos_vect= np.array([x,y])
 
-        val = np.dot(cl_vect - prev_vect, pos_vect - cl_vect)
+        val = np.dot(cl_vect- prev_vect, pos_vect - cl_vect)
 
-        if val > 0:
-            closest_idx = (closest_idx + 1) % len(self.waypoints_2d)
+        if val>0:
+            closest_idx = (closest_idx+1)% len(self.waypoints_2d)
         return closest_idx
 
-    def publish_waypoints(self, closest_idx):
-        lane = Lane()
-        lane.header = self.base_waypoints.header
-        # No buffer overrun, end is auto-truncated at end in Python
-        lane.waypoints = self.base_waypoints.waypoints[closest_idx : closest_idx + LOOKAHEAD_WPS]
-        # According to "4. WAypoint Updater Node (Partial)" final_waypoints_pub's type is styx_msgs/Lane
-        self.final_waypoints_pub.publish(lane)
+    def publish_waypoints(self):
+        final_lane = self.generate_lane()
+        self.final_waypoints_pub.publish(final_lane)
+
+    def generate_lane(self):
+        lane = lane()
+
+        closest_idx= self.get_closest_waypoint_idx()
+        farthest_idx = closest_idx+ LOOKAHEAD_WPS
+
+        if self.stopline_wp_idx == -1 or (self.stopline_wp_idx >= farthest_idx):
+            lane.waypoints = base_waypoints
+        else:
+            lane.waypoints = self.decelerate_waypoints(base_waypoints,closest_idx)
```

Comment:
`base_waypoints` is used without declaration here. Looks  
`self.base_waypoints` was renamed to `self.base_lane` (or maybe I did the reverse?), we need to use one or the other throughout.